Add --capture-ratio and --stats-only CLI args

DeeDeeG opened the pull request:
Adds some more (optional) CLI args that can be used, `--capture-ratio` and `--stats-only`:

```
% python3 ./sim.py --help
usage: sim.py [-h] [--presentmon-file PRESENTMON_FILE] [--capture-ratio CAPTURE_RATIO] [--stats-only]

Simulate OBS capture & compositing

options:
  -h, --help            show this help message and exit
  --presentmon-file PRESENTMON_FILE, --pmf PRESENTMON_FILE
                        use specified PresentMon capture file as pframe source
  --capture-ratio CAPTURE_RATIO, --cr CAPTURE_RATIO
                        capture no more than [this ratio] * [OBS FPS] times per second, loosely speaking (set to 0 for no limit)
  --stats-only, --silent, -s
                        print only the statistics, not the presented or captured frame info
```

---

`--stats-only` output currently looks like this:

```
% python3 ./sim.py --pmf ~/Downloads/pmcaps-TR/PresentMon-TR-Ultra-Uncapped-Gameplay-100-120FPS-ish-2024-01-30T170850.csv -s              
Data from: '/Users/user/Downloads/pmcaps-TR/PresentMon-TR-Ultra-Uncapped-Gameplay-100-120FPS-ish-2024-01-30T170850.csv'
Composite rate 60.0fps



===== STATS =====
Presented frames: 13982
Captured frames: 13605 (6406 unused)
Composited/output frames: 7199

Frame number gaps: 2.00 avg, 0 min, 4 max, 0.43 stddev
Frame time gaps: 17.255 avg, 5.610 min, 33.164 max, 3.294 stddev

```

---

I think the `--capture-ratio` is the more useful one, since one can always do `sim.py | grep stddev` to just grab the stats...

But offering them both, since I had them locally.

Open to requests for changes, review feedback, or feel free to edit the branch directly or land the changes however you like, of course. (If wanting these changes in the first place!)

DeeDeeG committed:
add --capture-ratio CLI arg

Adjusts how frequently the capture limit
allows a presented game frame to be captured.

A higher ratio allows more captures/second,
a lower ratio allows fewer captures/second,
as a multiple of the composite FPS.

Set this to 0 for no limit. Default is 2x the composite FPS.

DeeDeeG committed:
add --stats-only CLI arg

Useful when trying to grab stats without the terminal scrolling a lot.

Especially useful for repeated runs with slightly different settings,
or running the same settings against different capture csv files, etc.

Not recommended for initial/exploratory investigations, since it hides
a lot of under-the-hood details of what made those metrics happen.

Off by default.

alinsavix commented on sim.py:
```diff
@@ -185,7 +198,12 @@ def main(argv: List[str]) -> int:
+    elif args.capture_ratio == 0:
+        gc = GameCapture(0)
+    else:
+        gc = GameCapture(obs.composite_interval_ms / args.capture_ratio)
```
I would actually make the default for the argument be `2`, and then you can get rid of the special case for `None` -- since really, there's not actually a special case there, not specifying this argument means "use the default", and we can just... specify the default.

DeeDeeG replied:
Wasn't sure if there'd be a use for detecting this pref is explicitly set vs using the default, some time in the future maybe, but I don't know of a specific use for that, and certainly for now it just makes this if-else needlessly complicated.

alinsavix commented on sim.py:
```diff
@@ -217,9 +241,9 @@ def main(argv: List[str]) -> int:
         else:
             dispstr = frame.disposition.name
-        print(f"pframe {frame.present_frame} @ {frame.present_t_ms:0.3f}ms, {dispstr}")
+        frame_detail_print(f"pframe {frame.present_frame} @ {frame.present_t_ms:0.3f}ms, {dispstr}")
```
In this case, rather than moving the conditional logic to a function, I would just wrap this loop in one big `if not args.stats_only` so that we don't have to perform the loop _at all_, rather than still going through the loop a few thousand times (and making a function call each time) and then throwing away the output at the last moment.

DeeDeeG replied:
Ah, I saw the "Output/Composited Frames" section produced data used for the final stats, but I didn't notice this "Presented Frames" section only makes data for itself to use / is self-contained.

alinsavix reviewed (changes requested):
I have a few minor nits, but I would be more than happy to merge once those are resolved.

Thanks for the contribution!

DeeDeeG committed:
set default for --capture-ratio to "2"

Per review feedback

DeeDeeG committed:
do less work in stats-only mode

Per review feedback

DeeDeeG commented:
Sorry for the force push, I had an amended commit (5bedd928f183d8c1106ea2e5d19e428c387c6dc1) locally, but it looks like I had missed a line of `frame_detail_print()` in what I pushed to this PR's branch originally (b6d9de296ec45a113ad580b2e01ef05219357ee2). (Comparison: [`b6d9de2..5bedd92`](https://github.com/DeeDeeG/capturesim-fork/compare/b6d9de296ec45a113ad580b2e01ef05219357ee2..5bedd928f183d8c1106ea2e5d19e428c387c6dc1))

(If you dislike the force push, I can restore it and do a fixup commit instead... Would have to force push again to do that, but it'd mean the exact commit review comments were made based on would be in here again.)

And then the two latest commits are addressing the review feedback.

alinsavix commented:
> Sorry for the force push

In most projects I've seen, force pushes tend to be preferred -- certainly on any large project (which, admittedly, this is not) the preference tends to be "one commit per logical change" because there's really not that much benefit to retaining the winding path of small steps on the way to that change (and in most cases commits will be squashed anyhow). And generally the easiest way to arrive there is ... force pushes! When it's your own branch, anyhow. "Public" branches, not so much.
